### ISSUE TITLE
Formatting indentation for blocks

### ISSUE THREAD

**Issue by RobertWalterU3d** [eclipse/xtext#1807]
Imagine the following rules
```
Block:
	'block' name=ID ':'
	BEGIN
		blockStatements+=BlockStatement*
	END
	'endblock'
;

terminal BEGIN: 'synthetic:BEGIN';
terminal END: 'synthetic:END';
```
I tried many things in order to format that code so that every statement inside the block is indented, but I cannot make it work. I tried `block.interior[indent]`, which I would have thought should do the trick.

**Comment by miklossy** [comment]
Maybe the solution described here

https://www.eclipse.org/forums/index.php/t/1103866/

Is also applicabe to your use case as well.

**Comment by tivervac** [comment]
@RobertWalterU3d You have to make sure that you've defined rules for the BEGIN and END as well for interior to work

**Comment by RobertWalterU3d** [comment]
So I tried following the suggested solution and adjusting t to my concrete syntax:
```
	def dispatch void format(Block block, extension IFormattableDocument document) {
		formatBlockHeader(block, document)
		val begin = block.regionFor.keyword(':')
		val end = block.regionFor.keyword('endblock')
			
		interior(begin, end)[indent]
		for (blockStmt: block.blockStatements) {
			blockStmt.format
			blockStmt.prepend[indent].append[newLine]
		}
	}
```
This adds a second indent if I invoke the formatter on a correctly formatted file. If one statement is missing the indent, the formatter un-indents the immediately following line.

Before formatting:
```
block something:
   statement1
statement2 <--this should be indented
   statement3
```

```
block something:
   statement1
statement2
statement3 <-- formatting unindents the next line
```

**Comment by RobertWalterU3d** [comment]
> @RobertWalterU3d You have to make sure that you've defined rules for the BEGIN and END as well for interior to work

How can I access BEGIN and END in the formatter?

**Comment by tivervac** [comment]
Can you try to add these two rules?
```
block.regionFor.keyword("synthetic:BEGIN").append[newLine]
block.regionFor.keyword("synthetic:END").prepend[newLine]
```

The `blockStmt.prepend[indent].append[newLine]` shouldn't be necessary, the indent will be applied by the `interior`, the newline can be done in `def dispatch void format(BlockStatement bs, extension IFormattableDocument document)`

**Comment by cdietrich** [comment]
am not sure about the intention.
if the model is not intent
to begin with it wont parse

**Comment by RobertWalterU3d** [comment]
> Can you try to add these two rules?

Is this what you have in mind?
```
def dispatch void format(Block block, extension IFormattableDocument document) {
		formatBlockHeader(block, document)
		val begin = block.regionFor.ruleCallTo(BEGINRule)
		val end = block.regionFor.ruleCallTo(ENDRule)
			
		block.regionFor.keyword("synthetic:BEGIN").append[newLine]
		block.regionFor.keyword("synthetic:END").prepend[newLine]

		interior(begin, end)[indent]
		for (blockStmt: block.blockStatements) {
			blockStmt.format
		}
	}
```

**Comment by cdietrich** [comment]
i dont thunk the formatter api was written with intention based grammar in mind and 
did you check if it creates regions for the synthic tokens at all?

**Comment by RobertWalterU3d** [comment]
I don't know what "intention based grammar" means.

**Comment by cdietrich** [comment]
terminal BEGIN: 'synthetic:BEGIN';
terminal END: 'synthetic:END';

**Comment by tivervac** [comment]
@RobertWalterU3d That is indeed what I meant. I see you want to indent between the ':' and 'endblock' though so try to add the same NL prepend and append rules for those and see what that does.

**Comment by cdietrich** [comment]
this will work only if there is no parse error.
otherwise endblockKeyword_6
wont be found

**Comment by RobertWalterU3d** [comment]
> this will work only if there is no parse error. [...] and if there is no parse error the file is already formatted regarding ws

yes, it looks like it, and that is my "problem" I'd like to solve

**Comment by cdietrich** [comment]
i have only seen people having problem with the formatter and indentation based languages
unfortunately moritz does not even answer any questions anymore

**Comment by szarnekow** [comment]
Given that you have a begin and end keywords for your block, it's most likely easier to provide validations for the concrete syntax.
Your parser will gracefully munch through badly indented code, the formatter rules can rely on a correct model and the user will see far better error messages. Feasible?

**Comment by RobertWalterU3d** [comment]
> Given that you have a begin and end keywords for your block, it's most likely easier to provide validations for the concrete syntax.
> Your parser will gracefully munch through badly indented code, the formatter rules can rely on a correct model and the user will see far better error messages. Feasible?

This would be my option as well, but unfortunately, the language is given, I have no option to redesign it at that point. Also, there are other "Block-like" elements in the language without an "endblock" token... I would not have designed it this way, exactly for the reason you mention.

**Comment by tivervac** [comment]
Indeed the formatter only works correctly if there are no syntax errors in the input.

> there are other "Block-like" elements in the language without an "endblock" token.

You can try something like this
```
interior(start, block.blockStatements.last?.nextSemanticRegion)[indent]
```
